include/wccp/Wccp.h

```diff
@@ -121,7 +124,7 @@ class ServiceGroup : public ServiceConstants
   static uint8_t const RESERVED = 50;
 
   /// Number of ports in component (defined by protocol).
-  static size_t const N_PORTS = 8;
+  static constexpr size_t const N_PORTS = 8;
```

Comment:
Pretty sure this change has no practical effect, even hypothetically.

Reply:
The effect is you don't have to define this in a cpp file.

Reply:
> The effect is you don't have to define this in a cpp file.

If that were true, we'd have gotten link failures prior to this change.  You must be thinking of Objective-C or Swift or something.

Reply:
Objectively and swiftly so in this case:

```
[256/602] Linking CXX shared library src/wccp/libwccp.dylib
FAILED: src/wccp/libwccp.dylib 
: && /usr/bin/clang++ -g -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk -dynamiclib -Wl,-headerpad_max_install_names  -o src/wccp/libwccp.dylib -install_name @rpath/libwccp.dylib src/wccp/CMakeFiles/wccp.dir/WccpConfig.cc.o src/wccp/CMakeFiles/wccp.dir/WccpEndPoint.cc.o src/wccp/CMakeFiles/wccp.dir/WccpMsg.cc.o src/wccp/CMakeFiles/wccp.dir/WccpStatic.cc.o  -Wl,-rpath,/Users/cmcfarlen/projects/oss/trafficserver-cmake/cmake-build-debug/src/tscore -Wl,-rpath,/Users/cmcfarlen/projects/oss/trafficserver-cmake/cmake-build-debug/src/tscpp/util -Wl,-rpath,/Users/cmcfarlen/projects/oss/trafficserver-cmake/cmake-build-debug/lib/swoc -Wl,-rpath,/Users/cmcfarlen/projects/oss/trafficserver-cmake/cmake-build-debug/lib/yamlcpp  /opt/homebrew/Cellar/openssl@1.1/1.1.1t/lib/libcrypto.dylib  src/tscore/libtscore.dylib  /opt/homebrew/Cellar/openssl@1.1/1.1.1t/lib/libssl.dylib  /Library/Developer/CommandLineTools/SDKs/MacOSX13.3.sdk/usr/lib/libpcre.tbd  -lresolv  src/tscpp/util/libtscpputil.dylib  lib/swoc/libswoc-1.5.1.dylib  lib/yamlcpp/libyaml-cppd.0.7.0.dylib  /opt/homebrew/Cellar/openssl@1.1/1.1.1t/lib/libcrypto.dylib  /opt/homebrew/lib/libhwloc.dylib && :
Undefined symbols for architecture arm64:
  "wccp::ServiceGroup::N_PORTS", referenced from:
      (anonymous namespace)::Note_Svc_Ports_Too_Many(swoc::_1_5_1::Errata&, int, int) in WccpConfig.cc.o
ld: symbol(s) not found for architecture arm64
```

```
$ git diff
diff --git a/include/wccp/Wccp.h b/include/wccp/Wccp.h
index aa156c09d..abbec8eb1 100644
--- a/include/wccp/Wccp.h
+++ b/include/wccp/Wccp.h
@@ -124,7 +124,7 @@ public:
   static uint8_t const RESERVED = 50;

   /// Number of ports in component (defined by protocol).
-  static constexpr size_t const N_PORTS = 8;
+  static size_t const N_PORTS = 8;
```

Reply:
Hmm, why don't we have the same problem for HTTP or RESERVED?

Reply:
My hunch is N_PORTS is captured in the arguments to `errata.notes` in `Note_Svc_Ports_Too_Many`.  HTTP is unused and RESERVED is likely optimized away.

Reply:
So, why does the compiler complain about j and k here but not i?
```
wkaras /tmp/xx
$ cat x.cc
struct S
{
  static const int i{666};
  static const int j;
  static int k;
};

volatile int i;

int main()
{
  i = S::i;
  i = S::j;
  i = S::k;

  return 0;
}
wkaras /tmp/xx
$ gcc -O0 -Wall -Wextra -pedantic -std=c++17 x.cc
/usr/bin/ld: /tmp/ccGXX6cg.o: in function `main':
x.cc:(.text+0x14): undefined reference to `S::j'
/usr/bin/ld: x.cc:(.text+0x20): undefined reference to `S::k'
collect2: error: ld returned 1 exit status
wkaras /tmp/xx
$ 
```

Reply:
New theory:  the code's taking the address of N_PORTS.  Apparently `constexpr` is equivalent to `inline const`.  But `constexpr const` only serves Alan's  OCD.
```
wkaras /tmp/xx
$ cat y.cc
struct S
{
  static const int i{666};
  static constexpr int j{666};
  static constexpr const int k{666};
  static inline const int m{666};
};

const int * volatile p;

int main()
{
  p = &S::i;
  p = &S::j;
  p = &S::k;
  p = &S::m;

  return 0;
}
wkaras /tmp/xx
$ gcc -O0 -Wall -Wextra -pedantic -std=c++17 y.cc
/usr/bin/ld: /tmp/ccabFFfF.o: in function `main':
y.cc:(.text+0xb): undefined reference to `S::i'
collect2: error: ld returned 1 exit status
wkaras /tmp/xx
$
```

Reply:
I'm not going to worry about this because all of WCCP is going to be deprecated and removed. I think I added `constexpr` because I was blindly adding it to various this.